### ISSUE TITLE
Runtime error when initiating screencasting

### ISSUE THREAD

**Issue by Isti115** [waycrate/xdg-desktop-portal-luminous#16]
The `Settings` part of the portal seems to be working perfectly, but when I try to initiate a `ScreenCast`, I get the following error:
```rs
2023-12-29T10:54:58.517785Z  INFO
  dispatch_message {
    msg=Msg {
      type: MethodCall,
      sender: UniqueName(Str(Borrowed(":1.15"))),
      path: ObjectPath("/org/freedesktop/portal/desktop"),
      iface: InterfaceName(Str(Borrowed("org.freedesktop.impl.portal.ScreenCast"))),
      member: MemberName(Str(Borrowed("CreateSession"))), body: Signature("oosa{sv}")
    }
  }:dispatch_method_call {
    msg=Msg {
      type: MethodCall,
      sender: UniqueName(Str(Borrowed(":1.15"))),
      path: ObjectPath("/org/freedesktop/portal/desktop"),
      iface: InterfaceName(Str(Borrowed("org.freedesktop.impl.portal.ScreenCast"))),
      member: MemberName(Str(Borrowed("CreateSession"))),
      body: Signature("oosa{sv}")
    }
  }:dispatch_method_call_try {
    msg=Msg {
      type: MethodCall,
      sender: UniqueName(Str(Borrowed(":1.15"))),
      path: ObjectPath("/org/freedesktop/portal/desktop"),
      iface: InterfaceName(Str(Borrowed("org.freedesktop.impl.portal.ScreenCast"))),
      member: MemberName(Str(Borrowed("CreateSession"))),
      body: Signature("oosa{sv}")
    }
  }:
  xdg_desktop_portal_luminous::screencast:
  Start shot:
    path :/org/freedesktop/portal/desktop/request/1_77/obs4,
    appid: thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Connect(NoWaylandLib)', src/screencast.rs:194:63
```

I am running the binary from the development shell provided through `flake.nix`, so I assume, that the Wayland library should be present, as it's among the `nativeBuildInputs`. Also, `wayshot` works without any issues.

**Comment by Decodetalkers** [comment]
line 194 is libwayshot init connection, and it says there is no waylandlib

**Comment by Decodetalkers** [comment]
so, I think that is there is no wayland_display for the service

**Comment by Decodetalkers** [comment]
you can try to start it from commandline not systemd service..

![image](https://github.com/waycrate/xdg-desktop-portal-luminous/assets/60290287/850a4b2d-c549-4b70-8ac7-b87ca67a6430)

it should need a WAYLAND_DISPLAY

**Comment by Isti115** [comment]
Thanks for helping me troubleshoot this!
The error message was in fact captured by running it in the terminal, but I also have my `WAYLAND_DISPLAY` imported into the systemd environment, so there has got to be another cause at play.

In case anyone is facing the same issue, the temporary workaround I came up with was using the `wlr` portal for screencasting and relying on the luminous portal for the `accent-color` and `cholor-scheme` settings only.
`~/.config/xdg-desktop-portal/portals.conf`:
```ini
[preferred]
default=wlr
org.freedesktop.impl.portal.Settings=luminous
```

**Comment by Shinyzenith** [comment]
> Thanks for helping me troubleshoot this! The error message was in fact captured by running it in the terminal, but I also have my `WAYLAND_DISPLAY` imported into the systemd environment, so there has got to be another cause at play.
> 
> In case anyone is facing the same issue, the temporary workaround I came up with was using the `wlr` portal for screencasting and relying on the luminous portal for the `accent-color` and `cholor-scheme` settings only. `~/.config/xdg-desktop-portal/portals.conf`:
> 
> ```ini
> [preferred]
> default=wlr
> org.freedesktop.impl.portal.Settings=luminous
> ```

This is really interesting. Just to be clear, I am a Nix newbie and I copied the flake.nix from another project...maybe my flake code is the real issue here?

Are you sure luminous portal is receiving the WAYLAND_DISPLAY properly? Because wayshot and luminous use the same code for the backend ( libwayshot ). If wayshot works, there's no reason why luminous shouldn't.

**Comment by Isti115** [comment]
> Can you provide your distribution and compositor?

Of course, it's `NixOS 23.11` and `river 0.2.6`, but I tried under Sway as well and got the same result. I'll make a build that prints the environmental variables just to be sure that WAYLAND_DISPLAY is present and correct for the portal, but I would be very surprised if it wasn't, since as I've said, I have tried it from a terminal and have imported (and verified via `systemctl --user show-environment`) it into the systemd environment as well.

**Comment by Shinyzenith** [comment]
Hm still sounds like it's worth a shot. Please let me know!

**Comment by Arun-Singh-1242** [comment]
hm still sounds like it's wort a shot. please let me know!

**Comment by johnnywalker** [comment]
Thanks @elde-n! I took your solution a little farther and added a nixpkgs overlay with a wrapper for `xdg-desktop-portal-luminous`:

```nix
(final: prev: {
  xdg-desktop-portal-luminous = prev.xdg-desktop-portal-luminous.overrideAttrs (attrs: {
    nativeBuildInputs = attrs.nativeBuildInputs ++ [prev.makeWrapper];
    # allow `wayland-backend` create to find `libwayland-egl.so`
    postFixup = ''
      wrapProgram "$out/libexec/xdg-desktop-portal-luminous" \
                  --suffix LD_LIBRARY_PATH : "${prev.lib.strings.makeLibraryPath [prev.wayland]}"
    '';
  });
})
```
